File: frontend/scripts/audit-translations.mjs
#!/usr/bin/env node

/**
 * Translation Audit Script
 *
 * Checks translation files for:
 * 1. Missing keys between uk and en
 * 2. Placeholder values (TODO, TBD, ...)
 * 3. Ukrainian pluralization correctness
 * 4. Empty values
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const LOCALES_DIR = path.resolve(__dirname, '../public/locales');
const LANGUAGES = ['uk', 'en'];
const PLACEHOLDER_PATTERNS = [
  /TODO/i,
  /TBD/i,
  /\.\.\./,
  /xxx/i,
  /fixme/i,
  /placeholder/i
];

const REQUIRED_UK_PLURAL_SUFFIXES = ['_one', '_few', '_many'];
const REQUIRED_EN_PLURAL_SUFFIXES = ['_one', '_other'];

class TranslationAuditor {
  constructor() {
    this.issues = [];
    this.stats = {
      totalFiles: 0,
      totalKeys: { uk: 0, en: 0 },
      missingKeys: 0,
      placeholders: 0,
      emptyValues: 0,
      pluralIssues: 0
    };
  }

  /**
   * Load JSON file
   */
  loadJSON(filePath) {
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      return JSON.parse(content);
    } catch (error) {
      this.addIssue('error', `Failed to load ${filePath}: ${error.message}`);
      return null;
    }
  }

  /**
   * Get all translation file names
   */
  getTranslationFiles(lang) {
    const langDir = path.join(LOCALES_DIR, lang);
    if (!fs.existsSync(langDir)) {
      return [];
    }
    return fs.readdirSync(langDir).filter(f => f.endsWith('.json'));
  }

  /**
   * Flatten nested object to dot notation keys
   */
  flattenKeys(obj, prefix = '') {
    let keys = [];
    for (const [key, value] of Object.entries(obj)) {
      const fullKey = prefix ? `${prefix}.${key}` : key;
      if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        keys = keys.concat(this.flattenKeys(value, fullKey));
      } else {
        keys.push({ key: fullKey, value });
      }
    }
    return keys;
  }

  /**
   * Add issue to report
   */
  addIssue(level, message, file = null) {
    this.issues.push({ level, file, message });
  }

  /**
   * Check for placeholder values
   */
  checkPlaceholders(file, lang, entries) {
    const placeholders = [];

    for (const { key, value } of entries) {
      if (typeof value !== 'string') continue;

      for (const pattern of PLACEHOLDER_PATTERNS) {
        if (pattern.test(value)) {
          placeholders.push({ key, value, pattern: pattern.source });
          this.stats.placeholders++;
        }
      }
    }

    if (placeholders.length > 0) {
      this.addIssue(
        'warning',
        `Found ${placeholders.length} placeholder values:\n` +
        placeholders.map(p => `  - ${p.key}: "${p.value}"`).join('\n'),
        `${lang}/${file}`
      );
    }
  }

  /**
   * Check for empty values
   */
  checkEmptyValues(file, lang, entries) {
    const empty = entries.filter(({ value }) =>
      typeof value === 'string' && value.trim() === ''
    );

    if (empty.length > 0) {
      this.stats.emptyValues += empty.length;
      this.addIssue(
        'error',
        `Found ${empty.length} empty values:\n` +
        empty.map(e => `  - ${e.key}`).join('\n'),
        `${lang}/${file}`
      );
    }
  }

  /**
   * Check Ukrainian pluralization
   */
  checkUkrainianPlurals(file, data) {
    const pluralsSection = data.plurals;
    if (!pluralsSection) return;

    const keys = Object.keys(pluralsSection);
    const pluralGroups = {};

    // Group plural keys by base name
    for (const key of keys) {
      const match = key.match(/^(.+?)_(one|few|many)$/);
      if (match) {
        const [, base, form] = match;
        if (!pluralGroups[base]) pluralGroups[base] = new Set();
        pluralGroups[base].add(form);
      }
    }

    // Check each group has all required forms
    for (const [base, forms] of Object.entries(pluralGroups)) {
      const missing = REQUIRED_UK_PLURAL_SUFFIXES
        .map(s => s.replace('_', ''))
        .filter(form => !forms.has(form));

      if (missing.length > 0) {
        this.stats.pluralIssues++;
        this.addIssue(
          'error',
          `Plural "${base}" missing forms: ${missing.join(', ')}`,
          `uk/${file}`
        );
      }
    }
  }

  /**
   * Check English pluralization
   */
  checkEnglishPlurals(file, data) {
    const pluralsSection = data.plurals;
    if (!pluralsSection) return;

    const keys = Object.keys(pluralsSection);
    const pluralGroups = {};

    // Group plural keys by base name
    for (const key of keys) {
      const match = key.match(/^(.+?)_(one|other)$/);
      if (match) {
        const [, base, form] = match;
        if (!pluralGroups[base]) pluralGroups[base] = new Set();
        pluralGroups[base].add(form);
      }
    }

    // Check each group has all required forms
    for (const [base, forms] of Object.entries(pluralGroups)) {
      const missing = REQUIRED_EN_PLURAL_SUFFIXES
        .map(s => s.replace('_', ''))
        .filter(form => !forms.has(form));

      if (missing.length > 0) {
        this.stats.pluralIssues++;
        this.addIssue(
          'error',
          `Plural "${base}" missing forms: ${missing.join(', ')}`,
          `en/${file}`
        );
      }
    }
  }

  /**
   * Compare keys between languages
   */
  compareKeys(file, ukEntries, enEntries) {
    const ukKeys = new Set(ukEntries.map(e => e.key));
    const enKeys = new Set(enEntries.map(e => e.key));

    const missingInEn = [...ukKeys].filter(k => !enKeys.has(k));
    const missingInUk = [...enKeys].filter(k => !ukKeys.has(k));

    if (missingInEn.length > 0) {
      this.stats.missingKeys += missingInEn.length;
      this.addIssue(
        'error',
        `Missing ${missingInEn.length} keys in EN:\n` +
        missingInEn.map(k => `  - ${k}`).join('\n'),
        `en/${file}`
      );
    }

    if (missingInUk.length > 0) {
      this.stats.missingKeys += missingInUk.length;
      this.addIssue(
        'error',
        `Missing ${missingInUk.length} keys in UK:\n` +
        missingInUk.map(k => `  - ${k}`).join('\n'),
        `uk/${file}`
      );
    }
  }

  /**
   * Audit a single file pair
   */
  auditFile(file) {
    const ukPath = path.join(LOCALES_DIR, 'uk', file);
    const enPath = path.join(LOCALES_DIR, 'en', file);

    const ukData = this.loadJSON(ukPath);
    const enData = this.loadJSON(enPath);

    if (!ukData || !enData) return;

    const ukEntries = this.flattenKeys(ukData);
    const enEntries = this.flattenKeys(enData);

    this.stats.totalKeys.uk += ukEntries.length;
    this.stats.totalKeys.en += enEntries.length;

    // Check for missing keys
    this.compareKeys(file, ukEntries, enEntries);

    // Check for placeholders
    this.checkPlaceholders(file, 'uk', ukEntries);
    this.checkPlaceholders(file, 'en', enEntries);

    // Check for empty values
    this.checkEmptyValues(file, 'uk', ukEntries);
    this.checkEmptyValues(file, 'en', enEntries);

    // Check pluralization
    this.checkUkrainianPlurals(file, ukData);
    this.checkEnglishPlurals(file, enData);
  }

  /**
   * Run full audit
   */
  run() {
    console.log('🔍 Starting translation audit...\n');

    const ukFiles = this.getTranslationFiles('uk');
    const enFiles = this.getTranslationFiles('en');

    // Check for file discrepancies
    const allFiles = new Set([...ukFiles, ...enFiles]);
    const missingInEn = ukFiles.filter(f => !enFiles.includes(f));
    const missingInUk = enFiles.filter(f => !ukFiles.includes(f));

    if (missingInEn.length > 0) {
      this.addIssue(
        'error',
        `Files missing in EN: ${missingInEn.join(', ')}`
      );
    }

    if (missingInUk.length > 0) {
      this.addIssue(
        'error',
        `Files missing in UK: ${missingInUk.join(', ')}`
      );
    }

    // Audit each file
    for (const file of allFiles) {
      if (ukFiles.includes(file) && enFiles.includes(file)) {
        this.auditFile(file);
        this.stats.totalFiles++;
      }
    }

    this.printReport();
  }

  /**
   * Print audit report
   */
  printReport() {
    console.log('═══════════════════════════════════════════════════════');
    console.log('📊 TRANSLATION AUDIT REPORT');
    console.log('═══════════════════════════════════════════════════════\n');

    // Statistics
    console.log('📈 STATISTICS:');
    console.log(`   Files audited: ${this.stats.totalFiles}`);
    console.log(`   Total keys (UK): ${this.stats.totalKeys.uk}`);
    console.log(`   Total keys (EN): ${this.stats.totalKeys.en}`);
    console.log('');

    // Issues summary
    const errors = this.issues.filter(i => i.level === 'error').length;
    const warnings = this.issues.filter(i => i.level === 'warning').length;

    console.log('🚨 ISSUES SUMMARY:');
    console.log(`   Errors: ${errors}`);
    console.log(`   Warnings: ${warnings}`);
    console.log(`   - Missing keys: ${this.stats.missingKeys}`);
    console.log(`   - Placeholder values: ${this.stats.placeholders}`);
    console.log(`   - Empty values: ${this.stats.emptyValues}`);
    console.log(`   - Pluralization issues: ${this.stats.pluralIssues}`);
    console.log('');

    // Detailed issues
    if (this.issues.length > 0) {
      console.log('📋 DETAILED ISSUES:\n');

      const errorIssues = this.issues.filter(i => i.level === 'error');
      const warningIssues = this.issues.filter(i => i.level === 'warning');

      if (errorIssues.length > 0) {
        console.log('❌ ERRORS:');
        for (const issue of errorIssues) {
          if (issue.file) {
            console.log(`\n   📄 ${issue.file}`);
          }
          console.log(`   ${issue.message}`);
        }
        console.log('');
      }

      if (warningIssues.length > 0) {
        console.log('⚠️  WARNINGS:');
        for (const issue of warningIssues) {
          if (issue.file) {
            console.log(`\n   📄 ${issue.file}`);
          }
          console.log(`   ${issue.message}`);
        }
        console.log('');
      }
    } else {
      console.log('✅ NO ISSUES FOUND!\n');
    }

    console.log('═══════════════════════════════════════════════════════');

    // Exit with error code if issues found
    if (errors > 0) {
      console.log('\n❌ Audit failed with errors\n');
      process.exit(1);
    } else if (warnings > 0) {
      console.log('\n⚠️  Audit completed with warnings\n');
      process.exit(0);
    } else {
      console.log('\n✅ Audit passed successfully\n');
      process.exit(0);
    }
  }
}

// Run audit
const auditor = new TranslationAuditor();
auditor.run();

File: audit_translations.py
#!/usr/bin/env python3
"""
Translation Audit Script

Checks translation files for:
1. Missing keys between uk and en
2. Placeholder values (TODO, TBD, ...)
3. Ukrainian pluralization correctness
4. Empty values
"""

import json
import re
from pathlib import Path
from typing import Dict, List, Set, Tuple

LOCALES_DIR = Path(__file__).parent / "frontend" / "public" / "locales"
LANGUAGES = ["uk", "en"]
PLACEHOLDER_PATTERNS = [
    re.compile(r"TODO", re.IGNORECASE),
    re.compile(r"TBD", re.IGNORECASE),
    re.compile(r"\.\.\."),
    re.compile(r"xxx", re.IGNORECASE),
    re.compile(r"fixme", re.IGNORECASE),
    re.compile(r"placeholder", re.IGNORECASE),
]

REQUIRED_UK_PLURAL_SUFFIXES = ["_one", "_few", "_many"]
REQUIRED_EN_PLURAL_SUFFIXES = ["_one", "_other"]


class TranslationAuditor:
    def __init__(self):
        self.issues: List[Dict] = []
        self.stats = {
            "total_files": 0,
            "total_keys": {"uk": 0, "en": 0},
            "missing_keys": 0,
            "placeholders": 0,
            "empty_values": 0,
            "plural_issues": 0,
        }

    def load_json(self, file_path: Path) -> Dict | None:
        """Load JSON file"""
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception as e:
            self.add_issue("error", f"Failed to load {file_path}: {e}")
            return None

    def flatten_keys(self, obj: Dict, prefix: str = "") -> List[Tuple[str, str]]:
        """Flatten nested object to dot notation keys"""
        keys = []
        for key, value in obj.items():
            full_key = f"{prefix}.{key}" if prefix else key
            if isinstance(value, dict):
                keys.extend(self.flatten_keys(value, full_key))
            else:
                keys.append((full_key, value))
        return keys

    def add_issue(self, level: str, message: str, file: str | None = None):
        """Add issue to report"""
        self.issues.append({"level": level, "file": file, "message": message})

    def check_placeholders(self, file: str, lang: str, entries: List[Tuple[str, str]]):
        """Check for placeholder values"""
        placeholders = []

        for key, value in entries:
            if not isinstance(value, str):
                continue

            for pattern in PLACEHOLDER_PATTERNS:
                if pattern.search(value):
                    placeholders.append(
                        {"key": key, "value": value, "pattern": pattern.pattern}
                    )
                    self.stats["placeholders"] += 1
                    break

        if placeholders:
            msg = f"Found {len(placeholders)} placeholder values:\n"
            msg += "\n".join(f'  - {p["key"]}: "{p["value"]}"' for p in placeholders)
            self.add_issue("warning", msg, f"{lang}/{file}")

    def check_empty_values(self, file: str, lang: str, entries: List[Tuple[str, str]]):
        """Check for empty values"""
        empty = [(k, v) for k, v in entries if isinstance(v, str) and not v.strip()]

        if empty:
            self.stats["empty_values"] += len(empty)
            msg = f"Found {len(empty)} empty values:\n"
            msg += "\n".join(f"  - {k}" for k, _ in empty)
            self.add_issue("error", msg, f"{lang}/{file}")

    def check_ukrainian_plurals(self, file: str, data: Dict):
        """Check Ukrainian pluralization"""
        plurals_section = data.get("plurals")
        if not plurals_section:
            return

        keys = list(plurals_section.keys())
        plural_groups: Dict[str, Set[str]] = {}

        # Group plural keys by base name
        for key in keys:
            match = re.match(r"^(.+?)_(one|few|many)$", key)
            if match:
                base, form = match.groups()
                if base not in plural_groups:
                    plural_groups[base] = set()
                plural_groups[base].add(form)

        # Check each group has all required forms
        for base, forms in plural_groups.items():
            required = {s.replace("_", "") for s in REQUIRED_UK_PLURAL_SUFFIXES}
            missing = required - forms

            if missing:
                self.stats["plural_issues"] += 1
                self.add_issue(
                    "error",
                    f'Plural "{base}" missing forms: {", ".join(sorted(missing))}',
                    f"uk/{file}",
                )

    def check_english_plurals(self, file: str, data: Dict):
        """Check English pluralization"""
        plurals_section = data.get("plurals")
        if not plurals_section:
            return

        keys = list(plurals_section.keys())
        plural_groups: Dict[str, Set[str]] = {}

        # Group plural keys by base name
        for key in keys:
            match = re.match(r"^(.+?)_(one|other)$", key)
            if match:
                base, form = match.groups()
                if base not in plural_groups:
                    plural_groups[base] = set()
                plural_groups[base].add(form)

        # Check each group has all required forms
        for base, forms in plural_groups.items():
            required = {s.replace("_", "") for s in REQUIRED_EN_PLURAL_SUFFIXES}
            missing = required - forms

            if missing:
                self.stats["plural_issues"] += 1
                self.add_issue(
                    "error",
                    f'Plural "{base}" missing forms: {", ".join(sorted(missing))}',
                    f"en/{file}",
                )

    def compare_keys(
        self, file: str, uk_entries: List[Tuple], en_entries: List[Tuple]
    ):
        """Compare keys between languages"""
        uk_keys = {k for k, _ in uk_entries}
        en_keys = {k for k, _ in en_entries}

        missing_in_en = uk_keys - en_keys
        missing_in_uk = en_keys - uk_keys

        if missing_in_en:
            self.stats["missing_keys"] += len(missing_in_en)
            msg = f"Missing {len(missing_in_en)} keys in EN:\n"
            msg += "\n".join(f"  - {k}" for k in sorted(missing_in_en))
            self.add_issue("error", msg, f"en/{file}")

        if missing_in_uk:
            self.stats["missing_keys"] += len(missing_in_uk)
            msg = f"Missing {len(missing_in_uk)} keys in UK:\n"
            msg += "\n".join(f"  - {k}" for k in sorted(missing_in_uk))
            self.add_issue("error", msg, f"uk/{file}")

    def audit_file(self, file: str):
        """Audit a single file pair"""
        uk_path = LOCALES_DIR / "uk" / file
        en_path = LOCALES_DIR / "en" / file

        uk_data = self.load_json(uk_path)
        en_data = self.load_json(en_path)

        if not uk_data or not en_data:
            return

        uk_entries = self.flatten_keys(uk_data)
        en_entries = self.flatten_keys(en_data)

        self.stats["total_keys"]["uk"] += len(uk_entries)
        self.stats["total_keys"]["en"] += len(en_entries)

        # Check for missing keys
        self.compare_keys(file, uk_entries, en_entries)

        # Check for placeholders
        self.check_placeholders(file, "uk", uk_entries)
        self.check_placeholders(file, "en", en_entries)

        # Check for empty values
        self.check_empty_values(file, "uk", uk_entries)
        self.check_empty_values(file, "en", en_entries)

        # Check pluralization
        self.check_ukrainian_plurals(file, uk_data)
        self.check_english_plurals(file, en_data)

    def run(self):
        """Run full audit"""
        print("🔍 Starting translation audit...\n")

        uk_dir = LOCALES_DIR / "uk"
        en_dir = LOCALES_DIR / "en"

        uk_files = {f.name for f in uk_dir.glob("*.json")}
        en_files = {f.name for f in en_dir.glob("*.json")}

        # Check for file discrepancies
        all_files = uk_files | en_files
        missing_in_en = uk_files - en_files
        missing_in_uk = en_files - uk_files

        if missing_in_en:
            self.add_issue("error", f"Files missing in EN: {', '.join(missing_in_en)}")

        if missing_in_uk:
            self.add_issue("error", f"Files missing in UK: {', '.join(missing_in_uk)}")

        # Audit each file
        for file in sorted(all_files):
            if file in uk_files and file in en_files:
                self.audit_file(file)
                self.stats["total_files"] += 1

        self.print_report()

    def print_report(self):
        """Print audit report"""
        print("═══════════════════════════════════════════════════════")
        print("📊 TRANSLATION AUDIT REPORT")
        print("═══════════════════════════════════════════════════════\n")

        # Statistics
        print("📈 STATISTICS:")
        print(f'   Files audited: {self.stats["total_files"]}')
        print(f'   Total keys (UK): {self.stats["total_keys"]["uk"]}')
        print(f'   Total keys (EN): {self.stats["total_keys"]["en"]}')
        print()

        # Issues summary
        errors = sum(1 for i in self.issues if i["level"] == "error")
        warnings = sum(1 for i in self.issues if i["level"] == "warning")

        print("🚨 ISSUES SUMMARY:")
        print(f"   Errors: {errors}")
        print(f"   Warnings: {warnings}")
        print(f'   - Missing keys: {self.stats["missing_keys"]}')
        print(f'   - Placeholder values: {self.stats["placeholders"]}')
        print(f'   - Empty values: {self.stats["empty_values"]}')
        print(f'   - Pluralization issues: {self.stats["plural_issues"]}')
        print()

        # Detailed issues
        if self.issues:
            print("📋 DETAILED ISSUES:\n")

            error_issues = [i for i in self.issues if i["level"] == "error"]
            warning_issues = [i for i in self.issues if i["level"] == "warning"]

            if error_issues:
                print("❌ ERRORS:")
                for issue in error_issues:
                    if issue["file"]:
                        print(f'\n   📄 {issue["file"]}')
                    print(f'   {issue["message"]}')
                print()

            if warning_issues:
                print("⚠️  WARNINGS:")
                for issue in warning_issues:
                    if issue["file"]:
                        print(f'\n   📄 {issue["file"]}')
                    print(f'   {issue["message"]}')
                print()
        else:
            print("✅ NO ISSUES FOUND!\n")

        print("═══════════════════════════════════════════════════════")

        # Summary
        if errors > 0:
            print("\n❌ Audit failed with errors\n")
            exit(1)
        elif warnings > 0:
            print("\n⚠️  Audit completed with warnings\n")
        else:
            print("\n✅ Audit passed successfully\n")


if __name__ == "__main__":
    auditor = TranslationAuditor()
    auditor.run()
